File: tableau/Tableau.java
package tableau;

import java.util.ArrayList;

import deplacement.Deplacement;
import joueur.Joueur;
import scan.Scan;

public class Tableau
{
	private Boolean droite;
	private Boolean gauche;
	private Boolean haut;
	private Boolean bas;

	private Boolean porteDroite;
	private Boolean porteGauche;
	private Boolean porteHaut;
	private Boolean porteBas;
	private Levier levier;

	private int X;
	private int Y;
	
	
	private static ArrayList<Tableau> tableaux = new ArrayList<>();
	
	
	public Tableau(int X,  int Y)
	{
		this.setDroite(false);
		this.setGauche(false);
		this.setHaut(false);
		this.setBas(false);
		this.setPorteDroite(false);
		this.setPorteGauche(false);
		this.setPorteHaut(false);
		this.setPorteBas(false);
		this.setLevier(null);
		this.setX(X);
		this.setY(Y);
		tableaux.add(this);
		
	}

	public Boolean hasDroite()
	{
		try {
			getTableau(getX() + 1, getY());
			} catch (Exception e) {
			setDroite(true);
		}

		return droite;
	}

	
	public void setDroite(Boolean droite)
	{
		this.droite = droite;
	}

	public Tableau addDroite()
	{
		setDroite(true);
		return this;
	}
	public Tableau toDroite()
	{
		Joueur.get().setAncienTableauY(getX());
		return getTableau((getX() + 1), (getY()));
		
	}
	public void setPorteDroite(Boolean porteDroite) {
		this.porteDroite = porteDroite;
		this.droite = porteDroite;
	}
	public Tableau addPorteDroite() {
		setDroite(true);
		setPorteDroite(true);
		return this;
	}
	public Boolean hasPorteDroite() {
		return porteDroite;
	}
	
	public Boolean hasGauche()
	{
		try {
			getTableau(getX() - 1, getY());
			} catch (Exception e) {
				setGauche(true);
		}
			
		return gauche;
	}

	public void setGauche(Boolean gauche)
	{
		this.gauche = gauche;
	}
	public Tableau addGauche()
	{
		setGauche(true);
		return this;
	}
	public Tableau toGauche()
	{
		Joueur.get().setAncienTableauY(getX());
		return getTableau((getX() - 1), (getY()));
		
	}
	public void setPorteGauche(Boolean porteGauche) {
		this.porteGauche = porteGauche;
		this.gauche = porteGauche;
	}
	public Tableau addPorteGauche() {
		setGauche(true);
		setPorteGauche(true);
		return this;
	}
	public Boolean hasPorteGauche() {
		return porteGauche;
	}
	
	
	public Boolean hasHaut()
	{
		try {
			getTableau(getX(), getY() + 1);
		} catch (Exception e) {
			setHaut(true);
		}
		return haut;
	}

	public void setHaut(Boolean haut)
	{
		this.haut = haut;
	}
	public Tableau addHaut()
	{
		setHaut(true);
		return this;
	}
	public Tableau toHaut()
	{
		Joueur.get().setAncienTableauY(getY());
		return getTableau(getX(), (getY() + 1));
		
	}
	public void setPorteHaut(Boolean porteHaut) {
		this.porteHaut = porteHaut;
	}
	public Tableau addPorteHaut() {
		setHaut(true);
		setPorteHaut(true);
		return this;
	}
	public Boolean hasPorteHaut() {
		return porteHaut;
	}
	
	public Boolean hasBas()
	{
		try {
			getTableau(getX(), getY() - 1);
		} catch (Exception e) {
			setBas(true);
		}

		return bas;
	}

	public void setBas(Boolean bas)
	{
		this.bas = bas;
	}
	public Tableau addBas()
	{
		setBas(true);
		return this;
	}
	public Tableau toBas()
	{
		Joueur.get().setAncienTableauY(getY());
		return getTableau(getX(), (getY() - 1));
		
	}
	public void setPorteBas(Boolean porteBas) {
		this.porteBas = porteBas;
		this.bas = porteBas;
	}
	public Tableau addPorteBas() {
		setBas(true);
		setPorteBas(true);
		return this;
	}

	public Boolean hasPorteBas() {
		return porteBas;
	}
	
	public int getX()
	{
		return X;
	}

	public void setX(int X)
	{
		this.X = X;
	}
	
	public int getY()
	{
		return Y;
	}

	public void setY(int Y)
	{
		this.Y = Y;
	}

	
	public static ArrayList<Tableau> getTableaux()
	{
		return tableaux;
	}
	
	public static Tableau getTableau(int X, int Y)
	{	
		return getTableaux().stream().filter(t -> t.getX() == X && t.getY() == Y).findFirst().get();
	}

	public void evenement()
	{
		
		if(getX() == 3 && getY() == 0)
		{
			System.out.println("Bravo vous avez gagné ! ");
			System.exit(0);
			
		}
		else if(getX() == 0 && getY() == 0)
		{
			System.out.println("Vous vous trouvez au spawn.");
		}
		else if(this instanceof Tableau)
		{
			System.out.println("Vous êtes dans une plaine vide.");
		}
		else
		{
			System.out.println("Vous êtes dans le neant."); //Cas impossible
		}

		if(this.levier != null)
		{
			this.useLevier();
		}
		Deplacement.deplacement(this);
		
	}

	public Tableau setLevier(Levier levier)
	{
		this.levier = levier;
		return this;
	}

	public void useLevier()
	{
		Tableau tableau = Tableau.getTableau(levier.getX(), levier.getY());
		System.out.println("Vous trouvez un levier.\nVoulez vous l'utilisez ?");
		System.out.println("0 - Non");
		System.out.println("1 - Oui");

		boolean reponse = Scan.scannerBoolean();
		
	
		if(reponse)
		{
			if(!levier.isUsed()) // on/off levier 
			{
				System.out.println("Le levier est maintenant activé.");
				if (levier.getDroite()) {
					tableau.setPorteDroite(false);
				}
				if (levier.getGauche()) {
					tableau.setPorteGauche(false);
				}
				if (levier.getHaut()) {
					tableau.setPorteHaut(false);
				}
				if (levier.getBas()) {
					tableau.setPorteBas(false);
				}
				levier.setUsed(true);
			}
			else
			{
				System.out.println("Le levier est maintenant désactivé.");
				if (levier.getDroite()) {
					tableau.setPorteDroite(true);
				}
				if (levier.getGauche()) {
					tableau.setPorteGauche(true);
				}
				if (levier.getHaut()) {
					tableau.setPorteHaut(true);
				}
				if (levier.getBas()) {
					tableau.setPorteBas(true);
				}
				levier.setUsed(false);
			}
		}
	}
}

File: Cs/tableau/Tableau.cs
using System;
using System.Collections.Generic;
using Deplacement;
using Joueur;
using Scan;

namespace TableauNamespace
{
    public class Tableau
    {
        private bool droite;
        private bool gauche;
        private bool haut;
        private bool bas;

        private bool porteDroite;
        private bool porteGauche;
        private bool porteHaut;
        private bool porteBas;
        private Levier levier;

        private int x;
        private int y;

        private static List<Tableau> tableaux = new List<Tableau>();

        public Tableau(int x, int y)
        {
            this.droite = false;
            this.gauche = false;
            this.haut = false;
            this.bas = false;
            this.porteDroite = false;
            this.porteGauche = false;
            this.porteHaut = false;
            this.porteBas = false;
            this.levier = null;
            this.x = x;
            this.y = y;
            tableaux.Add(this);
        }

        public bool HasDroite()
        {
            try
            {
                GetTableau(x + 1, y);
            }
            catch (Exception)
            {
                droite = true;
            }
            return droite;
        }

        public void SetDroite(bool droite)
        {
            this.droite = droite;
        }

        public Tableau AddDroite()
        {
            SetDroite(true);
            return this;
        }

        public Tableau ToDroite()
        {
            Joueur.Get().SetAncienTableauY(x);
            return GetTableau(x + 1, y);
        }

        public void SetPorteDroite(bool porteDroite)
        {
            this.porteDroite = porteDroite;
            this.droite = porteDroite;
        }

        public Tableau AddPorteDroite()
        {
            SetDroite(true);
            SetPorteDroite(true);
            return this;
        }

        public bool HasPorteDroite()
        {
            return porteDroite;
        }

        public bool HasGauche()
        {
            try
            {
                GetTableau(x - 1, y);
            }
            catch (Exception)
            {
                gauche = true;
            }
            return gauche;
        }

        public void SetGauche(bool gauche)
        {
            this.gauche = gauche;
        }

        public Tableau AddGauche()
        {
            SetGauche(true);
            return this;
        }

        public Tableau ToGauche()
        {
            Joueur.Get().SetAncienTableauY(x);
            return GetTableau(x - 1, y);
        }

        public void SetPorteGauche(bool porteGauche)
        {
            this.porteGauche = porteGauche;
            this.gauche = porteGauche;
        }

        public Tableau AddPorteGauche()
        {
            SetGauche(true);
            SetPorteGauche(true);
            return this;
        }

        public bool HasPorteGauche()
        {
            return porteGauche;
        }

        public bool HasHaut()
        {
            try
            {
                GetTableau(x, y + 1);
            }
            catch (Exception)
            {
                haut = true;
            }
            return haut;
        }

        public void SetHaut(bool haut)
        {
            this.haut = haut;
        }

        public Tableau AddHaut()
        {
            SetHaut(true);
            return this;
        }

        public Tableau ToHaut()
        {
            Joueur.Get().SetAncienTableauY(y);
            return GetTableau(x, y + 1);
        }

        public void SetPorteHaut(bool porteHaut)
        {
            this.porteHaut = porteHaut;
        }

        public Tableau AddPorteHaut()
        {
            SetHaut(true);
            SetPorteHaut(true);
            return this;
        }

        public bool HasPorteHaut()
        {
            return porteHaut;
        }

        public bool HasBas()
        {
            try
            {
                GetTableau(x, y - 1);
            }
            catch (Exception)
            {
                bas = true;
            }
            return bas;
        }

        public void SetBas(bool bas)
        {
            this.bas = bas;
        }

        public Tableau AddBas()
        {
            SetBas(true);
            return this;
        }

        public Tableau ToBas()
        {
            Joueur.Get().SetAncienTableauY(y);
            return GetTableau(x, y - 1);
        }

        public void SetPorteBas(bool porteBas)
        {
            this.porteBas = porteBas;
            this.bas = porteBas;
        }

        public Tableau AddPorteBas()
        {
            SetBas(true);
            SetPorteBas(true);
            return this;
        }

        public bool HasPorteBas()
        {
            return porteBas;
        }

        public int GetX()
        {
            return x;
        }

        public void SetX(int x)
        {
            this.x = x;
        }

        public int GetY()
        {
            return y;
        }

        public void SetY(int y)
        {
            this.y = y;
        }

        public static List<Tableau> GetTableaux()
        {
            return tableaux;
        }

        public static Tableau GetTableau(int x, int y)
        {
            return GetTableaux().Find(t => t.GetX() == x && t.GetY() == y) 
                   ?? throw new Exception("Tableau not found");
        }

        public void Evenement()
        {
            if (x == 3 && y == 0)
            {
                Console.WriteLine("Bravo vous avez gagné !");
                Environment.Exit(0);
            }
            else if (x == 0 && y == 0)
            {
                Console.WriteLine("Vous vous trouvez au spawn.");
            }
            else if (this is Tableau)
            {
                Console.WriteLine("Vous êtes dans une plaine vide.");
            }
            else
            {
                Console.WriteLine("Vous êtes dans le néant."); // Cas impossible
            }

            if (levier != null)
            {
                UseLevier();
            }

            Deplacement.Deplacement(this);
        }

        public Tableau SetLevier(Levier levier)
        {
            this.levier = levier;
            return this;
        }

        public void UseLevier()
        {
            Tableau tableau = GetTableau(levier.GetX(), levier.GetY());
            Console.WriteLine("Vous trouvez un levier.\nVoulez-vous l'utiliser ?");
            Console.WriteLine("0 - Non");
            Console.WriteLine("1 - Oui");

            bool reponse = Scan.ScannerBoolean();

            if (reponse)
            {
                if (!levier.IsUsed()) // on/off levier 
                {
                    Console.WriteLine("Le levier est maintenant activé.");
                    if (levier.GetDroite())
                    {
                        tableau.SetPorteDroite(false);
                    }
                    if (levier.GetGauche())
                    {
                        tableau.SetPorteGauche(false);
                    }
                    if (levier.GetHaut())
                    {
                        tableau.SetPorteHaut(false);
                    }
                    if (levier.GetBas())
                    {
                        tableau.SetPorteBas(false);
                    }
                    levier.SetUsed(true);
                }
                else
                {
                    Console.WriteLine("Le levier est maintenant désactivé.");
                    if (levier.GetDroite())
                    {
                        tableau.SetPorteDroite(true);
                    }
                    if (levier.GetGauche())
                    {
                        tableau.SetPorteGauche(true);
                    }
                    if (levier.GetHaut())
                    {
                        tableau.SetPorteHaut(true);
                    }
                    if (levier.GetBas())
                    {
                        tableau.SetPorteBas(true);
                    }
                    levier.SetUsed(false);
                }
            }
        }
    }
}
